### ISSUE TITLE
Provide a set of utility methods to help handling URLs

### ISSUE THREAD

**Issue by bobylito** [algolia/algoliasearch-helper-js#203]
As we continue the development of the instantsearch.js, we need to provide a set of utility methods on the helper to make saving and restoring a search state easily (see [related issue in the instant search project](https://github.com/algolia/instantsearch.js/issues/17)). We will need to provide 3 methods on the helper : 

 - one that creates a plain javascript object with only the specified subset of parameters from the current search state. This one could be implemented as specific behavior of the current getState method.
 - two that will serialize and deserialize a given subset of parameters and transform them into a query string form

The format to specify the subset is : 
```js
[
  // Any parameters of the SearchParameters can be chosen. Most likely to be chosen : 
  "query",
  "page",
  "hitsPerPage",
  // Also the index, as it can be changed for sorting purpose but is not in the SearchParameters
  "index",
  // Facets can be set through the usage of different widgets. But the user don't know much about that, only that it is based on an attribute
  "attribute:myAttribute",
  // For simplicity sake, we can choose to track all the facets.
  "attribute:*"
]
```

## API proposal

```js
var filters = ["numeric:price", "facet:category", "query"]
helper.getState(filters);
helper.getStateAsQueryString(filters, moreAttributes);
helper.setStateFromQueryString(queryString, filters);
```

Update : remove the queryString() factory method and use only plain methods

**Comment by pixelastic** [comment]
I feel it counter-intuitive that `helper.queryString()` returns an object and not a real query string. I would see more something like a `helper.exportStateToQueryString(filters)` and `helper.setStateFromQueryString(queryString)` instead

**Comment by vvo** [comment]
Great that we (YOU :D) are going to do this!

> helper.queryString().serialize(filters, moreAttributes);
What is `moreAttributes`?

> helper.queryString().deserialize(serializedString, filters);

Why do you have a first step being `.queryString()`? Will we be able to somehow use its result. If so what would it look like?

Also if `.queryString()` is a sort of first step then I thought it would accepts `.queryString(filters..)`?

As for the API proposal, what about:

```js
helper.getStateAsQueryString(filters);
helper.setStateFromQueryString(string, filters);
helper.getState(filters);
```

?

Maybe you need more than that for other use cases, you tell me.

> The format to specify the subset is :
> 
> [
>   // Same goes with numerics (because they are not actual facets)
>   "numeric:*",
>   "numeric:price"
> ]

Is this required to provide the `numeric: token` or could we just have `attributes: ['prices', 'category']` and we would magically map them to facets/disjunctiveFacets/hierarchicalFacets/numericFilters?

**Comment by bobylito** [comment]
I'm glad you had questions about queryString(). It's a way to modularize the code a little bit. This way we can avoid having yet again methods in the same big bag that has become the helper. Those methods start to become annoying as we have big and sometimes confusing name for them.

But it (querystring) is wrong in this current form, the functions don't expect a state so there is no need for it to be a function. But we could make use of the fact it is function as @vvo suggests, it will make it stateful and thus will require to save and share it between the components of the app. This  could be alright though (we already ask that from the user with the helper) and I think I like that :)

@vvo not sure that we can do too much magic here. We can know that an attribute is used as a facet because it is declared by the user, but can't really do that for the numerics. They can be described as disjunctive facets, and used as numeric filters...

**Comment by vvo** [comment]
> They can be described as disjunctive facets, and used as numeric filters...

When they are declared as facets or disjunctive facets and used as numeric filters then we know it's a numeric filter in the helper state (state.numericFilters?) so I thought we can do the magic.

Maybe you mean "when they are used both like regular facets AND as numeric filters" but I did not see such usage for now (did you?).

> But it (querystring) is wrong in this current form, the functions don't expect a state so there is no need for it to be a function.

Ok so you want a namespace inside the helper to store url related methods.

It can be done but not always easy to keep references (helper main instance) around automatically (on an implementation side).

What about `helper.state.getAsQueryString`, `helper.state.setFromQueryString` then?

> Those methods start to become annoying as we have big and sometimes confusing name for them.

I feel it's ok right now when I use the helper, the doc is pretty good to be ctrl+f and find the right method having the "state" token.

Even if we split inside another namespace we will still add methods and thus will have to somehow document and try to guide our users through them.

Let me know.

API DESIGN! :D

**Comment by vvo** [comment]
Also a comment of mine was badly formated:

> helper.queryString().serialize(filters, moreAttributes);

What is `moreAttributes`?

**Comment by bobylito** [comment]
> What is moreAttributes?

It's for the case the user has other attributes in the url. He can pass them as a way to consolidate every parameters into a single  query string.

> When they are declared as facets or disjunctive facets and used as numeric filters then we know it's a numeric filter in the helper state (state.numericFilters?) so I thought we can do the magic.

Not until we start to do refinement on them. I guess you are right because when we deserialize we will know it's numerics...

> I feel it's ok right now

I guess it also feels OK because you've been hacking on it for months now :p 

I think the modularization still makes more sense as : querystring() (I'm doing something with a querystring) serialize (I know I'm in the context of the querystring)

Also if we want to provide a lighter version of the module without the querystring serialization, it's easy to make querystring to throw an exception.

But the alternative works as well. Just wanted to improve and renew the way we pile up things around here.

**Comment by bobylito** [comment]
@vvo if we have facet, numeric and so on, we can have a more fine grained way to specify the attributes when using the `*`. For example : 
 - ['attribute:*'] will retrieve everything and easily make an ugly URL
 - whereas ['facet:*', 'numeric:price'] it more precise

But the users may not need that... :)

**Comment by vvo** [comment]
What do you think of "What about helper.state.getAsQueryString, helper.state.setFromQueryString then?" ?

**Comment by vvo** [comment]
> whereas ['facet:*', 'numeric:price'] it more precise

Yes I only fear we are adding "numeric:" parameter which does not exists in the helper right now and might be more confusing (same thing with attribute: I agree).

**Comment by bobylito** [comment]
I like `helper.state.getAsQueryString`, `helper.state.setFromQueryString` less but that's OK if you both like them more :) (does make the code and documentation easier)

**Comment by bobylito** [comment]
> Yes I only fear we are adding "numeric:" parameter which does not exists (as a parameter) in the helper right now and might be more confusing (same thing with attribute: I agree).

Yes sure, let's go for attributes.

**Comment by pixelastic** [comment]
I like `helper.state.getAsQueryString`, `helper.state.setFromQueryString` better than going through an intermediate object.

**Comment by bobylito** [comment]
@pixelastic Cool let's do that then (issue already updated :) )

**Comment by vvo** [comment]
Something we did not plan but will be necessary for some implementations: being able to prefix every parameter to avoid naming conflicts.

For example I could have a page=faq and it will conflict with page=0 from the search state.

For this case we can provide a `prefix: 'aa'` option something like that. Only here for reference for you @bobylito we talked about it

**Comment by pixelastic** [comment]
Why don't we wrap all our variables in only on variable of the queryString?

I mean, instead of `?page=faq&aa-page=0&aa-facets=price,brand&aa-whatever=foo,bar`, shouldn't we have `?page=faq&aa-url=[page:0,facets:price;brand,whatever:foo;bar]` (of course we need to find another way to encode our data).

My point is: to avoid naming conflicts with already existing variables in query strings, we should use only one, with a carefully crafted and unique name that encapsulate in a custom format all we need.

**Comment by vvo** [comment]
> My point is: to avoid naming conflicts with already existing variables in query strings, we should use only one

That's a good idea yes I like it. I hope that's feasible.

We can still use https://github.com/hapijs/qs, we will need to do (pseudo code) `aa-state=qs(state)`.

And then all we need to provide is the `aa-state` token as a parameter (that's the prefix).

 thoughts @bobylito?

**Comment by pixelastic** [comment]
![image](https://cloud.githubusercontent.com/assets/283419/9759021/954fe724-56ea-11e5-9c03-9c3d711720d8.png)

Because we'll need to urlencode the output of qs (a querystring) to pass it as one of the variables, right?

**Comment by vvo** [comment]
> Because we'll need to urlencode the output of qs (a querystring) to pass it as one of the variables, right?

I guess yes, we will see :D

**Comment by bobylito** [comment]
I prefer the flattest option without the state attribute because
 - it would mean to encode something that is already encoded and then make the qs way bigger
 - it's an edge case, and instantsearch.js is for solving the problems of the majority

:)

**Comment by pixelastic** [comment]
I'm not sure I get what you're saying, sorry. I'm not sure we're talking about the same thing.

We won't encode something already encoded, we'll just encode it once. And this will avoid repeating the `prefix` on each variable, so this might actually not make the qs bigger. Also not sure how this is an edge case, people will have all sorts of variables in their urls and we must not mess with them.

**Comment by vvo** [comment]
I had the discussion with @bobylito and yes @pixelastic what you proposed would look like this (both in the url bar and in copy pasted links):

`http://website.com/search.php?page=sales&aa-url=query%3Dhello%20man%26facetsRefinements%5B%5D%3Dred%26facetsRefinements%5B%5D%3Dblue` => because you must encode a queryString if you store it inside a queryString parameter.

While with the previous way it would look like this:

`http://website.com/search.php?page=sales&query=hello man&facetsRefinements[]=red&facetsRefinements[]=blue`

**Comment by pixelastic** [comment]
Nice example, makes the discussion much clearer. The second version is much more readable and I like it better, actually.

- We can add prefixes to each variable to avoid naming conflicts
- With aliases thrown in, it will be even smaller (`fr[]=red&fr[]=blue`)
- Could we even do `fr=[red,blue]` ?

The last point that we discussed with @bobylito was how to check for corrupted url (truncated link for example). With everything in the same variable we just have to check if we can decode it. With it split into several variables, we could have incomplete data.

**Comment by redox** [comment]
> With aliases thrown in, it will be even smaller (fr[]=red&fr[]=blue)

And with `arrayFormat: 'repeat'` it would even be like: `fr=red&fr=blue`

**Comment by vvo** [comment]
> The last point that we discussed with @bobylito was how to check for corrupted url (truncated link for example). With everything in the same variable we just have to check if we can decode it. With it split into several variables, we could have incomplete data.

I do not understand this, I think you are ok for the "no everything in a single query string param" but then I do not get your point here since it's in favor of everything inside a single query strnig param.

**Comment by bobylito** [comment]
This issue is fixed with #209
